main/lsp/request_dispatch.cc

```diff
@@ -21,8 +25,262 @@ class SorbetFenceTask final : public LSPTask {
+            }
+            evictedHashes[fref.id()] = move(globalStateHashes[fref.id()]);
+            globalStateHashes[fref.id()] = update.updatedFileHashes[i];
+            frefs.push_back(fref);
+        }
+    }
+
+    // Index changes in initialGS. pipeline::index sorts output by file id, but we need to reorder to match the order of
+    // other fields.
+    UnorderedMap<u2, int> fileToPos;
+    {
+        int i = -1;
+        for (auto fref : frefs) {
+            // We should have ensured before reaching here that there are no duplicates.
+            ENFORCE(!fileToPos.contains(fref.id()));
+            i++;
+            fileToPos[fref.id()] = i;
+        }
+    }
+
+    {
+        auto trees = pipeline::index(initialGS, frefs, config->opts, *emptyWorkers, kvstore);
+        initialGS->errorQueue->drainWithQueryResponses(); // Clear error queue; we don't care about errors here.
+        update.updatedFileIndexes.resize(trees.size());
+        for (auto &ast : trees) {
+            const int i = fileToPos[ast.file.id()];
+            update.updatedFileIndexes[i] = move(ast);
+        }
+    }
+
+    auto runningSlowPath = initialGS->epochManager->getStatus();
+    if (runningSlowPath.slowPathRunning) {
+        // A cancelable slow path is currently running. Before running deepCopy(), check if we can cancel -- we might be
+        // able to avoid it.
+        // Invariant: `pendingTypecheckUpdates` should contain the edits currently being typechecked on the slow path.
+        // runningSlowPath.epoch should be in the interval (pendingTypecheckUpdates.epoch - editCount,
+        // pendingTypecheckUpdates.epoch]
+        ENFORCE(runningSlowPath.epoch <= pendingTypecheckUpdates.epoch);
+        ENFORCE(runningSlowPath.epoch > (pendingTypecheckUpdates.epoch - pendingTypecheckUpdates.editCount));
+        auto merged = update.copy();
+        merged.mergeOlder(pendingTypecheckUpdates);
+        auto mergedEvictions = mergeEvictions(pendingTypecheckEvictedStateHashes, evictedHashes);
+        merged.canTakeFastPath = canTakeFastPath(*initialGS, *config, globalStateHashes, merged, mergedEvictions);
+        // Cancel if old + new takes fast path, or if the new update will take the slow path anyway.
+        if ((merged.canTakeFastPath || !update.canTakeFastPath) &&
+            initialGS->epochManager->tryCancelSlowPath(merged.epoch)) {
+            // Cancelation succeeded! Use `merged` as the update.
+            update = move(merged);
+            update.canceledSlowPath = true;
+            evictedHashes = std::move(mergedEvictions);
+        }
+    }
+
+    ENFORCE(update.updatedFiles.size() == update.updatedFileHashes.size());
+    ENFORCE(update.updatedFiles.size() == update.updatedFileIndexes.size());
+
+    // Completely replace `pendingTypecheckUpdates` if this was a slow path update.
+    if (!update.canTakeFastPath) {
+        update.updatedGS = initialGS->deepCopy();
+        pendingTypecheckUpdates = update.copy();
```

Comment:
`pendingTypecheckUpdates` contains all of the edits that were last run on the slow path.

If the slow path finishes, it may be stale -- but that's OK. We only ever use it if the `epochManager` says we are running a slow path, and we succeed in canceling it.